### PR TITLE
[JN-1112] create core family linkage models

### PR DESCRIPTION
**Pull request by connorlbark**
#### DESCRIPTION (include screenshots, and mobile screenshots for participant UX)

Make sure you agree with how I modeled / linked families and enrollees.

Note that `familyRelationship` is freetext - these relationships will be defined by study staff. I think it would be a can of worms to make it an enum.

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

- ensure study config can enable and disable family linkage (otherwise no functional behavior)

### CODE REVIEW

File: core/src/main/java/bio/terra/pearl/core/model/participant/RelationshipType.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package bio.terra.pearl.core.model.participant;
 
 public enum RelationshipType{
-    PROXY;
+    PROXY, FAMILY;
```
Comment (connorlbark) [inline core/src/main/java/bio/terra/pearl/core/model/participant/RelationshipType.java:4]: There are many places (afaik only in the UI) that treat _any_ enrollee relation as providing full access; fixing that is the subject of a future PR.

File: core/src/test/java/bio/terra/pearl/core/service/survey/SurveyTaskDispatcherTest.java [hunk 2]
```diff
@@ -250,7 +247,7 @@ public void testAssignOnSurveyEvent(TestInfo testInfo) {
         assertThat(participantTaskService.findByEnrolleeId(bundle.enrollee().getId()), hasSize(2));
         tasks = participantTaskService.findByEnrolleeId(bundle.enrollee().getId());
         // now the task should be added
-        assertThat(tasks.stream().map(ParticipantTask::getTargetStableId).toList(), contains("medForm", "followUp"));
+        assertThat(tasks.stream().map(ParticipantTask::getTargetStableId).toList(), containsInAnyOrder("medForm", "followUp"));
```
Comment (connorlbark) [inline core/src/test/java/bio/terra/pearl/core/service/survey/SurveyTaskDispatcherTest.java]: drive by flaky test fix

File: core/src/main/java/bio/terra/pearl/core/service/participant/FamilyService.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import bio.terra.pearl.core.dao.participant.FamilyDao;
+import bio.terra.pearl.core.model.participant.Family;
+import bio.terra.pearl.core.service.CrudService;
+import bio.terra.pearl.core.service.exception.internal.InternalServerException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class FamilyService extends CrudService<Family, FamilyDao> {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/participant/FamilyService.java]: This should extend DataAuditedService so we can track creation/deletion of families

File: core/src/main/java/bio/terra/pearl/core/service/participant/FamilyService.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            family.setShortcode(generateShortcode());
+        }
+        Family savedFamily = dao.create(family);
+        return savedFamily;
+    }
+
+    /**
+     * Generate a unique shortcode for a family. Operates exactly like the generateShortcode method in
+     * EnrolleeService.
+     */
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/participant/FamilyService.java]: let's pull this into a utlity method, and also let's make sure that Enrollee and Family shortcodes have the same length and allowed chars.   I think the right approach is to have a ShortcodeUtils  class with the LENGTH and ALLOWED_CHARS utility method that takes as an argument a `shortcodeAlreadyExists` String -> boolean function.   

Should we prefix Family shortcodes with `F_` or some similar prefix so they're distinguishable at a glance from Enrollee shortcodes?

**Review by devonbush — CHANGES_REQUESTED** [review]
Overall, agree with the approach.  The blocker is that this needs to be audited.  And we should do some thinking about the best way to handle the shortcodes

File: core/src/main/java/bio/terra/pearl/core/service/participant/ShortcodeUtilService.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import bio.terra.pearl.core.service.exception.internal.InternalServerException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+@Service
+public class ShortcodeUtilService {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/participant/ShortcodeUtilService.java]: Generally, something is either a utility or a service.  I think of utilities as non-Spring components that encapsulate pure business logic.  So I'd rename this to ShortcodeService to connote that this hits the database  and has some spring stuff injected into it

File: ui-participant/src/test-utils/test-proxy-environment.tsx [hunk 6]
```diff
@@ -1,5 +1,12 @@
...
-import { EnrolleeRelation, ParticipantUser, PortalParticipantUser } from '../api/api'
+import {
+  Enrollee,
+  Portal
+} from '@juniper/ui-core'
+import {
+  EnrolleeRelation,
+  ParticipantUser,
+  PortalParticipantUser
+} from '../api/api'
```
Comment (devonbush) [inline ui-participant/src/test-utils/test-proxy-environment.tsx]:
```suggestion
} from 'api/api'
```

**Review by devonbush — APPROVED** [review]
nice, and thanks for remembering to update the enrollee withdraw logic!

**Review by MatthewBemis — APPROVED** [review]
:shipit:

**Pull request merged**